halo2_proofs/src/dev/failure/emitter.rs

```diff
@@ -0,0 +1,164 @@
+        FailureLocation::OutsideRegion { row } => {
+            eprintln!("{}Cell layout at row {}:", prefix, row);
+            eprint!("{}  |Rotation|", prefix);
+            None
+        }
+    };
+
+    // Print the assigned cells, and their region offset or rotation.
+    for (column, cells) in columns {
+        let width = col_width(*cells);
+        eprint!(
+            "{}|",
+            padded(
+                ' ',
+                width,
+                &format!(
+                    "{}{}",
+                    match column.column_type {
+                        Any::Advice => "A",
+                        Any::Fixed => "F",
+                        Any::Instance => "I",
+                    },
+                    column.index,
+                )
+            )
+        );
+    }
+    eprintln!();
+    eprint!("{}  +--------+", prefix);
+    for cells in columns.values() {
+        eprint!("{}+", padded('-', col_width(*cells), ""));
+    }
+    eprintln!();
+    for (rotation, row) in layout {
+        eprint!(
+            "{}  |{}|",
+            prefix,
+            padded(' ', 8, &(offset.unwrap_or(0) + rotation).to_string())
+        );
+        for (col, cells) in columns {
+            let width = col_width(*cells);
+            eprint!(
+                "{}|",
+                padded(
+                    ' ',
+                    width,
+                    row.get(col).map(|s| s.as_str()).unwrap_or_default()
+                )
+            );
+        }
+        highlight_row(offset, *rotation);
+        eprintln!();
+    }
+}
+
+pub(super) fn expression_to_string<F: Field>(
+    expr: &Expression<F>,
+    layout: &BTreeMap<i32, BTreeMap<metadata::Column, String>>,
+) -> String {
+    expr.evaluate(
+        &util::format_value,
+        &|_| panic!("virtual selectors are removed during optimization"),
+        &|query, column, rotation| {
+            if let Some(label) = layout
+                .get(&rotation.0)
+                .and_then(|row| row.get(&(Any::Fixed, column).into()))
+            {
+                label.clone()
+            } else if rotation.0 == 0 {
+                // This is most likely a merged selector
+                format!("S{}", query)
+            } else {
+                // No idea how we'd get here...
+                format!("F{}@{}", column, rotation.0)
+            }
+        },
+        &|_, column, rotation| {
+            layout
+                .get(&rotation.0)
+                .unwrap()
+                .get(&(Any::Advice, column).into())
+                .unwrap()
+                .clone()
+        },
+        &|_, column, rotation| {
+            layout
+                .get(&rotation.0)
+                .unwrap()
+                .get(&(Any::Instance, column).into())
+                .unwrap()
+                .clone()
+        },
+        &|a| {
+            if a.contains(' ') {
+                format!("-({})", a)
+            } else {
+                format!("-{}", a)
```

Comment:
This doesn't contain space, so technically you could end up with `--x` instead of `-(-x)`. It probably doesn't matter; I can't see a case where the result would be ambiguous.